src/protobuf.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _PROTOBUF_H
+#define _PROTOBUF_H
+
+#include "arch.h"
+#include <string.h>
+#include <type_traits>
+
+typedef const u32 protobuf_t;
+static protobuf_t VARINT = 0;
+static protobuf_t LEN = 2;
+
+typedef u32 protobuf_index_t;
+
+// We assume the length of a nested field can be represented with 3 varint
+// bytes.
+const size_t nested_field_byte_count = 3;
+
+class ProtobufBuffer {
+private:
+  ProtobufBuffer *_parent_message;
+  char *_data;
```

Comment:
Maybe `unsigned char` to make less casts?
I know that FlightRecorder buffer uses `char`, so I don't mind either option.